File: Examples/Comparison/midpoint_root_compare.py
from copy import deepcopy
from io import StringIO
from time import perf_counter

from Bio import Phylo
from Bio.Align import MultipleSeqAlignment
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor


def build_sequences(sequence_count=60, width=160):
    template = ("ACGT" * ((width + 3) // 4))[:width]
    bases = "ACGT"
    records = []
    for sequence_index in range(1, sequence_count + 1):
        chars = list(template)
        for position in range(1, width + 1):
            if (sequence_index + position) % 13 == 0:
                chars[position - 1] = bases[(sequence_index + position - 2) % len(bases)]
            elif (sequence_index * position) % 29 == 0:
                chars[position - 1] = bases[(sequence_index + 2 * position - 3) % len(bases)]
        records.append(SeqRecord(Seq("".join(chars)), id=f"seq{sequence_index}", description="synthetic"))
    return records


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def build_tree():
    alignment = MultipleSeqAlignment(build_sequences())
    calculator = DistanceCalculator("identity")
    constructor = DistanceTreeConstructor()
    distance_matrix = calculator.get_distance(alignment)
    return constructor.nj(distance_matrix)


def main():
    tree = build_tree()

    def midpoint_root():
        rooted = deepcopy(tree)
        rooted.root_at_midpoint()
        return rooted

    midpoint_ms, rooted = time_call(midpoint_root)

    assert len(rooted.get_terminals()) == len(tree.get_terminals())

    print(f"python_midpoint_root_ms={midpoint_ms:.4f}")
    print(f"python_midpoint_root_ok={str(True).lower()}")


if __name__ == "__main__":
    main()

File: Examples/Comparison/midpoint_root_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using Random
using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function build_sequences(sequence_count::Integer=60, width::Integer=160)
    Random.seed!(1234)
    template = repeat("ACGT", cld(width, 4))[1:width]
    bases = ('A', 'C', 'G', 'T')
    sequences = String[]

    for sequence_index in 1:sequence_count
        chars = collect(template)
        for position in 1:width
            if mod(sequence_index + position, 13) == 0
                chars[position] = bases[mod(sequence_index + position - 2, length(bases)) + 1]
            elseif mod(sequence_index * position, 29) == 0
                chars[position] = bases[mod(sequence_index + 2 * position - 3, length(bases)) + 1]
            end
        end
        push!(sequences, String(chars))
    end

    return sequences
end

function build_tree()
    sequences = build_sequences()
    labels = ["seq$(index)" for index in eachindex(sequences)]
    distances = BioToolkit.distance_matrix(sequences; method=:hamming)
    return BioToolkit.neighbor_joining_tree(distances, labels)
end

function main()
    tree = build_tree()
    midpoint_ms, midpoint_tree = repeat_elapsed_ms(() -> BioToolkit.midpoint_root(tree), 100)

    @assert length(BioToolkit.get_terminals(midpoint_tree)) == length(BioToolkit.get_terminals(tree))
    @assert occursin("(", BioToolkit.write_newick(midpoint_tree))

    python_output = read(`conda run -n general --no-capture-output python $(joinpath(@__DIR__, "midpoint_root_compare.py"))`, String)
    python_midpoint_ms = parse(Float64, match(r"python_midpoint_root_ms=([0-9.]+)", python_output).captures[1])
    python_midpoint_ok = match(r"python_midpoint_root_ok=(true|false)", python_output).captures[1] == "true"

    @assert python_midpoint_ok

    println("Midpoint rooting benchmark")
    println("  sequence_count=60")
    println("  julia_midpoint_root_ms=$(round(midpoint_ms / 100; digits=4))")
    println("  python_midpoint_root_ms=$(round(python_midpoint_ms; digits=4))")
    print(python_output)
end

main()
